# Lesson 01 ... Welcome to pandas

We will start with importing some libraries we need and then play with some data to understand basic python commands. What data shall we work with? Well, let us pull down some data on criminal incidences that were reported.

First we install a particular library called `pandas` and in the command that follows, note that `pd` is just the alias that pandas assumes so that we can type `pd` and have all the `pandas` commands at our disposal.

In [21]:
import pandas as pd 
import numpy as np

The crime incident reports data are [available here]("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv") and span multiple years so we may end up working only with 2019 data but for now we proceed by gathering everything.

In the command below, the key part is `pd.read_csv()` and inside it is the URL for the comma-separated variable file. Once the file is downloaded by `pandas` we are saving it in python with the name `df` 

Note that data-sets, data-files are usually referred to as a `data-frame` in python and hence the alias of `df`.


In [22]:
df = pd.read_csv("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv")

/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (0,2,6,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Let us look at the first 5 rows of data to get a feel for the layout. The command is `.head(5)`

In [23]:
df.head(5)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP        OFFENSE_DESCRIPTION  \
0       TESTTEST2           423                NaN       ASSAULT - AGGRAVATED   
1       S97333701          3301                NaN             VERBAL DISPUTE   
2       S47513131          2647                NaN  THREATS TO DO BODILY HARM   
3       I92102201          3301                NaN             VERBAL DISPUTE   
4       I92097173          3115                NaN         INVESTIGATE PERSON   

   DISTRICT REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH  \
0  External                       0  2019-10-16 00:00:00  2019     10   
1        C6            915        0  2020-07-18 14:34:00  2020      7   
2       E18            530        0  2020-06-24 10:15:00  2020      6   
3       E13            583        0  2019-12-20 03:08:00  2019     12   
4       C11            355        0  2019-10-23 00:00:00  2019     10   

  DAY_OF_WEEK  HOUR UCR_PART          STREET        Lat       Long  \
0   Wednesday     0      NaN    RIVERVIEW DR        NaN        NaN   
1    Saturday    14      NaN  MARY BOYLE WAY  42.330813 -71.051368   
2   Wednesday    10      NaN    READVILLE ST  42.239491 -71.135954   
3      Friday     3      NaN          DAY ST  42.325122 -71.107779   
4   Wednesday     0      NaN       GIBSON ST  42.297555 -71.059709   

                      Location  
0     (0.00000000, 0.00000000)  
1  (42.33081300, -71.05136800)  
2  (42.23949100, -71.13595400)  
3  (42.32512200, -71.10777900)  
4  (42.29755500, -71.05970900)

What about the last 10 rows of the data?

In [24]:
df.tail(10)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP  \
515072       102095489          3115                NaN   
515073       102091671          2647                NaN   
515074       020224065          3018                NaN   
515075       020202856          2672                NaN   
515076       020063425          3114                NaN   
515077       020062356          3115                NaN   
515078       020054040          3501                NaN   
515079       020046400          1501                NaN   
515080       020038446          1501                NaN   
515081       020030892           540                NaN   

                                      OFFENSE_DESCRIPTION DISTRICT  \
515072                                 INVESTIGATE PERSON      E18   
515073                          THREATS TO DO BODILY HARM       B3   
515074                      SICK/INJURED/MEDICAL - POLICE       B2   
515075                                 BIOLOGICAL THREATS       B2   
515076                               INVESTIGATE PROPERTY       A7   
515077                                 INVESTIGATE PERSON      E18   
515078                                     MISSING PERSON      C11   
515079  WEAPON VIOLATION - CARRY/ POSSESSING/ SALE/ TR...       B2   
515080  WEAPON VIOLATION - CARRY/ POSSESSING/ SALE/ TR...       B2   
515081                              BURGLARY - COMMERICAL      C11   

       REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH DAY_OF_WEEK  \
515072            520        0  2019-11-25 16:30:00  2019     11      Monday   
515073            417        0  2019-11-12 12:00:00  2019     11     Tuesday   
515074            282        0  2020-03-19 07:30:00  2020      3    Thursday   
515075            282        0  2020-03-19 08:30:00  2020      3    Thursday   
515076             21        0  2020-09-01 00:00:00  2020      9     Tuesday   
515077            520        0  2020-08-28 18:39:00  2020      8      Friday   
515078                       0  2020-07-30 15:30:00  2020      7    Thursday   
515079            330        0  2020-07-02 01:38:00  2020      7    Thursday   
515080            300        0  2020-06-03 01:15:00  2020      6   Wednesday   
515081            380        0  2020-05-03 00:00:00  2020      5      Sunday   

        HOUR UCR_PART         STREET        Lat       Long  \
515072    16      NaN  HYDE PARK AVE  42.256215 -71.124019   
515073    12      NaN        MORA ST  42.282081 -71.073648   
515074     7      NaN  WASHINGTON ST  42.353272 -71.173738   
515075     8      NaN      WARREN ST  42.328234 -71.083289   
515076     0      NaN       PARIS ST  42.374426 -71.035278   
515077    18      NaN  HYDE PARK AVE  42.256215 -71.124019   
515078    15      NaN      GIBSON ST  42.297555 -71.059709   
515079     1      NaN    PASADENA RD  42.305760 -71.083771   
515080     1      NaN  WASHINGTON ST  42.323807 -71.089150   
515081     0      NaN  GALLIVAN BLVD  42.283700 -71.047761   

                           Location  
515072  (42.25621500, -71.12401900)  
515073  (42.28208100, -71.07364800)  
515074  (42.35327200, -71.17373800)  
515075  (42.32823400, -71.08328900)  
515076  (42.37442600, -71.03527800)  
515077  (42.25621500, -71.12401900)  
515078  (42.29755500, -71.05970900)  
515079  (42.30576000, -71.08377100)  
515080  (42.32380700, -71.08915000)  
515081  (42.28370000, -71.04776100)

Let us look at the contents of the data-frame ... 

| Column Name | Description |
| :--          | :--- |
| [incident_num] [varchar](20) NOT NULL, | Internal BPD report number |
| [offense_code][varchar](25) NULL,| Numerical code of offense description |
| [Offense_Code_Group_Description][varchar](80) NULL, | Internal categorization of [offense_description] |
| [Offense_Description][varchar](80) NULL, | Primary descriptor of incident |
| [district] [varchar](10) NULL,| What district the crime was reported in |
| [reporting_area] [varchar](10) NULL, | RA number associated with the where the crime was reported from. |
| [shooting][char] (1) NULL, | Indicated a shooting took place. |
| [occurred_on] [datetime2](7) NULL, | Earliest date and time the incident could have taken place |
| [UCR_Part] [varchar](25) NULL,| Universal Crime Reporting Part number (1,2, 3) |
| [street] [varchar](50) NULL,| Street name the incident took place |


Offense Codes are [available here](https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx)

We could also look at the offense codes by reading them in as a data-frame. This is an Excel file so we will have to switch to `.read_excel()`


In [25]:
offense_codes = pd.read_excel("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx")

In [26]:
print(offense_codes)

     CODE                                       NAME
0     612           LARCENY PURSE SNATCH - NO FORCE 
1     613                        LARCENY SHOPLIFTING
2     615    LARCENY THEFT OF MV PARTS & ACCESSORIES
3    1731                                     INCEST
4    3111                  LICENSE PREMISE VIOLATION
..    ...                                        ...
571  1806  DRUGS - CLASS B TRAFFICKING OVER 18 GRAMS
572  1807  DRUGS - CLASS D TRAFFICKING OVER 50 GRAMS
573  1610    HUMAN TRAFFICKING - COMMERCIAL SEX ACTS
574  2010                              HOME INVASION
575  1620  HUMAN TRAFFICKING - INVOLUNTARY SERVITUDE

[576 rows x 2 columns]


The next step would be to see how many data points we have, and what the minimum, maximum values, what is the average, etc. This can be done with `.describe()`

In [27]:
df.describe()

OFFENSE_CODE           YEAR          MONTH           HOUR  \
count  515082.000000  515082.000000  515082.000000  515082.000000   
mean     2333.275632    2017.542933       6.634194      13.079170   
std      1182.489822       1.543329       3.317964       6.347259   
min       111.000000    2015.000000       1.000000       0.000000   
25%      1102.000000    2016.000000       4.000000       9.000000   
50%      3005.000000    2018.000000       7.000000      14.000000   
75%      3201.000000    2019.000000       9.000000      18.000000   
max      3831.000000    2020.000000      12.000000      23.000000   

                 Lat           Long  
count  485909.000000  485909.000000  
mean       42.239043     -70.949353  
std         1.891645       3.060012  
min        -1.000000     -71.203312  
25%        42.296861     -71.097465  
50%        42.325029     -71.077723  
75%        42.348312     -71.062562  
max        42.395042       0.000000

By default the command will report the values with decimals but we may not want that. Decimals can be `rounded` or removed altogether as shown below.

In [28]:
df.describe().round(2)

OFFENSE_CODE       YEAR      MONTH       HOUR        Lat       Long
count     515082.00  515082.00  515082.00  515082.00  485909.00  485909.00
mean        2333.28    2017.54       6.63      13.08      42.24     -70.95
std         1182.49       1.54       3.32       6.35       1.89       3.06
min          111.00    2015.00       1.00       0.00      -1.00     -71.20
25%         1102.00    2016.00       4.00       9.00      42.30     -71.10
50%         3005.00    2018.00       7.00      14.00      42.33     -71.08
75%         3201.00    2019.00       9.00      18.00      42.35     -71.06
max         3831.00    2020.00      12.00      23.00      42.40       0.00

In [29]:
df.describe().round(0)

OFFENSE_CODE      YEAR     MONTH      HOUR       Lat      Long
count      515082.0  515082.0  515082.0  515082.0  485909.0  485909.0
mean         2333.0    2018.0       7.0      13.0      42.0     -71.0
std          1182.0       2.0       3.0       6.0       2.0       3.0
min           111.0    2015.0       1.0       0.0      -1.0     -71.0
25%          1102.0    2016.0       4.0       9.0      42.0     -71.0
50%          3005.0    2018.0       7.0      14.0      42.0     -71.0
75%          3201.0    2019.0       9.0      18.0      42.0     -71.0
max          3831.0    2020.0      12.0      23.0      42.0       0.0

Note a few things here. 

* We have a total of 515082 incidents logged. But the latitude and longitude are availale for no more than 485909 incidents. 


Say we want to restrict the dataframe just to 2020. How can we do that?

In [30]:
df20 = df[ df['YEAR'] == 2020 ]

Notice the sequence here `dataframe[ dataframe['column-name'] == somevalue ]` & pay attention to the double equal sign `==` which is a strict equality. 

In [31]:
df20.describe()

OFFENSE_CODE     YEAR         MONTH          HOUR           Lat  \
count  63733.000000  63733.0  63733.000000  63733.000000  62200.000000   
mean    2353.137323   2020.0      4.900554     12.923525     42.319872   
std     1182.670996      0.0      2.561463      6.566899      0.032339   
min      111.000000   2020.0      1.000000      0.000000     42.181845   
25%     1001.000000   2020.0      3.000000      9.000000     42.295353   
50%     3005.000000   2020.0      5.000000     14.000000     42.321918   
75%     3207.000000   2020.0      7.000000     18.000000     42.344561   
max     3831.000000   2020.0      9.000000     23.000000     42.395041   

               Long  
count  62200.000000  
mean     -71.084193  
std        0.030578  
min      -71.203312  
25%      -71.098579  
50%      -71.078444  
75%      -71.062000  
max      -70.953726

At this point we might be curious to know what types of offenses are most often reported? Before we that, however, let us also see how many unique values of OFFENSE_CODE are there

In [32]:
df20['OFFENSE_CODE'].nunique()

130

In [33]:
df20['OFFENSE_CODE'].value_counts()

3301    6234
3115    5494
801     3908
3005    3227
3831    2700
        ... 
1603       2
3203       1
2672       1
990        1
2628       1
Name: OFFENSE_CODE, Length: 130, dtype: int64

So code 3301 leads with 6234 reports in 2020, followed by code 3115, then 801, then 3005, and then 3831. Code 3005 is missing from their list so we have no idea what it is!! That is a crime in itself.   

In [34]:
# Just another way to accomplish the same thing but in a more complicated way.

df20.groupby('OFFENSE_CODE')['OFFENSE_CODE'].count().reset_index(name='count').sort_values(['count'], ascending = False) 

OFFENSE_CODE  count
109          3301   6234
95           3115   5494
23            801   3908
82           3005   3227
129          3831   2700
..            ...    ...
86           3016      2
26            990      1
106          3203      1
76           2672      1
65           2628      1

[130 rows x 2 columns]

Let us focus in on these verbal disputes. We will do so by creating a new dataframe that is only for OFFENSE_CODE 3301.

In [35]:
dfverbal = df20[ df20['OFFENSE_CODE'] == 3301 ]

In [36]:
# Now we see this dataframe just to check

dfverbal

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP OFFENSE_DESCRIPTION  \
1            S97333701          3301                NaN      VERBAL DISPUTE   
13           I20205924          3301                NaN      VERBAL DISPUTE   
305302       202062797          3301                NaN      VERBAL DISPUTE   
306053       202053448          3301                NaN      VERBAL DISPUTE   
426898       212014631          3301                NaN      VERBAL DISPUTE   
...                ...           ...                ...                 ...   
491035       202000014          3301                NaN      VERBAL DISPUTE   
491069       200359470          3301                NaN      VERBAL DISPUTE   
491101       200204118          3301                NaN      VERBAL DISPUTE   
491108       200157167          3301                NaN      VERBAL DISPUTE   
491114       200123335          3301                NaN      VERBAL DISPUTE   

       DISTRICT REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH  \
1            C6            915        0  2020-07-18 14:34:00  2020      7   
13           E5            822        0  2020-07-19 03:10:00  2020      7   
305302       E5                       0  2020-08-30 11:08:00  2020      8   
306053       D4            623        0  2020-07-28 13:32:00  2020      7   
426898       B3            943        0  2020-02-23 06:52:00  2020      2   
...         ...            ...      ...                  ...   ...    ...   
491035       D4            164        0  2020-01-01 00:30:00  2020      1   
491069       B2            316        0  2020-07-16 22:52:00  2020      7   
491101       A7            824        0  2020-04-28 19:10:00  2020      4   
491108      E13            572        0  2020-03-28 09:18:00  2020      3   
491114       B2                       0  2020-03-08 00:00:00  2020      3   

       DAY_OF_WEEK  HOUR UCR_PART            STREET        Lat       Long  \
1         Saturday    14      NaN    MARY BOYLE WAY  42.330813 -71.051368   
13          Sunday     3      NaN       EDGEMERE RD  42.258676 -71.151164   
305302      Sunday    11      NaN               NaN  42.293540 -71.122772   
306053     Tuesday    13      NaN  COMMONWEALTH AVE        NaN        NaN   
426898      Sunday     6      NaN      WOODBOLE AVE  42.277355 -71.079358   
...            ...   ...      ...               ...        ...        ...   
491035   Wednesday     0      NaN      HARRISON AVE  42.341362 -71.067076   
491069    Thursday    22      NaN      HOMESTEAD ST  42.312421 -71.091549   
491101     Tuesday    19      NaN        TRENTON ST  42.379684 -71.034286   
491108    Saturday     9      NaN      COLUMBUS AVE  42.313627 -71.095603   
491114      Sunday     0      NaN          DEVON ST  42.309277 -71.078687   

                           Location  
1       (42.33081300, -71.05136800)  
13      (42.25867600, -71.15116400)  
305302  (42.29354000, -71.12277200)  
306053     (0.00000000, 0.00000000)  
426898  (42.27735500, -71.07935800)  
...                             ...  
491035  (42.34136200, -71.06707600)  
491069  (42.31242100, -71.09154900)  
491101  (42.37968400, -71.03428600)  
491108  (42.31362700, -71.09560300)  
491114  (42.30927700, -71.07868700)  

[6234 rows x 17 columns]

Which days of the week have more verbal disputes?

In [37]:
dfverbal['DAY_OF_WEEK'].value_counts()

Friday       944
Saturday     939
Sunday       918
Thursday     878
Tuesday      866
Wednesday    857
Monday       832
Name: DAY_OF_WEEK, dtype: int64

Which hour, which streets have the most verbal disputes?

In [38]:
dfverbal['HOUR'].value_counts()

21    418
20    401
18    381
0     365
19    364
22    352
16    329
17    324
23    321
11    307
12    295
15    287
14    272
13    270
10    260
1     242
9     217
2     182
8     150
7     130
3     122
4      97
6      81
5      67
Name: HOUR, dtype: int64

In [39]:
dfverbal['STREET'].value_counts()

WASHINGTON ST     209
BLUE HILL AVE     113
COLUMBIA RD        91
CENTRE ST          84
DORCHESTER AVE     81
                 ... 
US-20               1
FLAHERTY WAY        1
DEVER ST            1
FAIRBANKS ST        1
DELL AVE            1
Name: STREET, Length: 1405, dtype: int64

What districts are the worst?

In [40]:
dfverbal['DISTRICT'].value_counts()

C11         1250
B3          1231
B2          1208
E18          476
D4           427
D14          296
E5           285
C6           275
E13          273
A7           265
A1           135
A15          103
External       3
Name: DISTRICT, dtype: int64

Lookup districts C11, B3, and B2 ... what areas are these?

# Practice Task 01

Pick another data-set from `data.boston.gov` and go through the same commands, picking some interesting element of the dataframe to explore